packages/postgres-database/src/simcore_postgres_database/utils_folders_v2.py

```diff
@@ -0,0 +1,916 @@
+
+
+class BaseAccessError(FoldersError):
+    pass
+
+
+class FolderNotFoundError(BaseAccessError):
+    msg_template = "no entry for folder_id={folder_id} found"
+
+
+class FolderNotSharedWithGidError(BaseAccessError):
+    msg_template = "folder_id={folder_id} was not shared with gid={gid}"
+
+
+class InsufficientPermissionsError(BaseAccessError):
+    msg_template = "could not find a parent for folder_id={folder_id} and gid={gid}, with permissions={permissions}"
+
+
+class BaseCreateFlderError(FoldersError):
+    pass
+
+
+class FolderAlreadyExistsError(BaseCreateFlderError):
+    msg_template = (
+        "A folder='{folder}' with parent='{parent}' for group='{gid}' already exists"
+    )
+
+
+class ParentFolderIsNotWritableError(BaseCreateFlderError):
+    msg_template = "Cannot create any sub-folders inside folder_id={parent_folder_id} since it is not writable for gid={gid}."
+
+
+class CouldNotCreateFolderError(BaseCreateFlderError):
+    msg_template = "Could not create folder='{folder}' and parent='{parent}'"
+
+
+class GroupIdDoesNotExistError(BaseCreateFlderError):
+    msg_template = "Provided group id '{gid}' does not exist "
+
+
+class BaseMoveFolderError(FoldersError):
+    pass
+
+
+class CannotMoveFolderSharedViaNonPrimaryGroupError(BaseMoveFolderError):
+    msg_template = (
+        "deltected group_type={group_type} for gid={gid} which is not allowed"
+    )
+
+
+class BaseAddProjectError(FoldersError):
+    pass
+
+
+class ProjectAlreadyExistsInFolderError(BaseAddProjectError):
+    msg_template = "project_id={project_id} in folder_id={folder_id} is already present"
+
+
+###
+### UTILS ACCESS LAYER
+###
+
+
+class FolderAccessRole(Enum):
+    """Used by the frontend to indicate a role in a simple manner"""
+
+    NO_ACCESS = 0
+    VIEWER = 1
+    EDITOR = 2
+    OWNER = 3
+
+
+class _FolderPermissions(TypedDict):
+    read: bool
+    write: bool
+    delete: bool
+
+
+def _make_permissions(
+    *, r: bool = False, w: bool = False, d: bool = False, description: str = ""
+) -> "_FolderPermissions":
+    _ = description
+    return _FolderPermissions(read=r, write=w, delete=d)
+
+
+def _only_true_permissions(permissions: _FolderPermissions) -> dict:
+    return {k: v for k, v in permissions.items() if v is True}
+
+
+def _or_reduce(x: _FolderPermissions, y: _FolderPermissions) -> _FolderPermissions:
+    return _FolderPermissions(
+        {
+            "read": x["read"] or y["read"],
+            "write": x["write"] or y["write"],
+            "delete": x["delete"] or y["delete"],
+        }
+    )
+
+
+def _or_dicts_list(dicts: Iterable[_FolderPermissions]) -> _FolderPermissions:
+    if not dicts:
+        return _make_permissions()
+    return reduce(_or_reduce, dicts)
+
+
+class _BasePermissions:
+    LIST_FOLDERS: ClassVar[_FolderPermissions] = _make_permissions(r=True)
+    # NOTE `LIST_PROJECTS` bypasses these access rights and always lists all projects in the fodler
```

Comment:
what does that mean in this context? list projects? you are talking about the webserver???